core/src/main/java/org/elasticsearch/index/percolator/QueryMetadataService.java

```diff
@@ -0,0 +1,151 @@
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + term.field(), term.bytes()));
+        } else if (query instanceof PhraseQuery) {
+            Term[] terms = ((PhraseQuery) query).getTerms();
+            if (terms.length == 0) {
+                return;
+            }
+
+            Term longestTerm = terms[0];
+            for (Term term : terms) {
+                if (longestTerm.bytes().length < term.bytes().length) {
+                    longestTerm = term;
+                }
+            }
+            queryTerms.add(new Term(QUERY_METADATA_FIELD_PREFIX + longestTerm.field(), longestTerm.bytes()));
+        } else if (query instanceof BooleanQuery) {
+            List<BooleanClause> clauses = ((BooleanQuery) query).clauses();
+            boolean noShouldClauses = true;
+            for (BooleanClause clause : clauses) {
+                if (clause.getOccur() == BooleanClause.Occur.SHOULD) {
+                    noShouldClauses = false;
+                    break;
+                }
+            }
+            if (noShouldClauses) {
+                List<Term> bestClause = null;
+                for (BooleanClause clause : clauses) {
+                    if (clause.isProhibited()) {
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+
+                    List<Term> temp = new ArrayList<>();
+                    extractQueryMetadata(clause.getQuery(), temp);
+                    if (bestClause == null) {
+                        bestClause = temp;
+                    } else {
+                        int currentSize = 0;
+                        for (Term term : bestClause) {
+                            currentSize += term.bytes().length;
+                        }
+                        int otherSize = 0;
+                        for (Term term : temp) {
+                            otherSize += term.bytes().length;
+                        }
+                        // keep the clause with longest terms, this likely to be rarest.
+                        if (otherSize > currentSize) {
+                            bestClause = temp;
+                        }
+                    }
+                }
+                if (bestClause != null) {
+                    queryTerms.addAll(bestClause);
+                }
+            } else {
+                for (BooleanClause clause : clauses) {
+                    if (clause.isProhibited()) {
+                        // we don't need to remember the things that do *not* match...
+                        continue;
+                    }
+                    extractQueryMetadata(clause.getQuery(), queryTerms);
+                }
+            }
+        } else if (query instanceof ConstantScoreQuery) {
+            Query wrappedQuery = ((ConstantScoreQuery) query).getQuery();
+            extractQueryMetadata(wrappedQuery, queryTerms);
+        } else {
```

Comment:
I wanted to do this, but the `TermsQuery` doesn't have methods to access the terms it encapsulates.

I'll add support for `BoostQuery`.